tests/components/eafm/test_sensor.py

```diff
@@ -0,0 +1,431 @@
+"""Tests for polling measures."""
+import datetime
+
+import aiohttp
+import pytest
+
+from homeassistant import config_entries
+from homeassistant.setup import async_setup_component
+import homeassistant.util.dt as dt_util
+
+from tests.common import MockConfigEntry, async_fire_time_changed
+
+DUMMY_REQUEST_INFO = aiohttp.client.RequestInfo(
+    url="http://example.com", method="GET", headers={}, real_url="http://example.com"
+)
+
+CONNECTION_EXCEPTIONS = [
+    aiohttp.ClientConnectionError("Mock connection error"),
+    aiohttp.ClientResponseError(DUMMY_REQUEST_INFO, [], message="Mock response error"),
+]
+
+
+async def async_setup_test_fixture(hass, mock_get_station, initial_value):
+    """Create a dummy config entry for testing polling."""
+    mock_get_station.return_value = initial_value
+
+    entry = MockConfigEntry(
+        version=1,
+        domain="eafm",
+        entry_id="VikingRecorder1234",
+        data={"station": "L1234"},
+        title="Viking Recorder",
+        connection_class=config_entries.CONN_CLASS_CLOUD_PUSH,
+    )
+    entry.add_to_hass(hass)
+
+    assert await async_setup_component(hass, "eafm", {})
+    assert entry.state == config_entries.ENTRY_STATE_LOADED
+    await hass.async_block_till_done()
+
+    async def poll(value):
+        mock_get_station.reset_mock(return_value=True, side_effect=True)
+
+        if isinstance(value, Exception):
+            mock_get_station.side_effect = value
+        else:
+            mock_get_station.return_value = value
+
+        next_update = dt_util.utcnow() + datetime.timedelta(60 * 15)
+        async_fire_time_changed(hass, next_update)
+        await hass.async_block_till_done()
+
+    return entry, poll
+
+
+async def test_reading_measures_not_list(hass, mock_get_station):
+    """
+    Test that a measure can be a dict not a list.
+
+    E.g. https://environment.data.gov.uk/flood-monitoring/id/stations/751110
+    """
+    _ = await async_setup_test_fixture(
+        hass,
+        mock_get_station,
+        {
+            "label": "My station",
+            "measures": {
+                "@id": "really-long-unique-id",
+                "label": "York Viking Recorder - level-stage-i-15_min----",
+                "qualifier": "Stage",
+                "parameterName": "Water Level",
+                "latestReading": {"value": 5},
+                "stationReference": "L1234",
+            },
+        },
+    )
+
+    state = hass.states.get("sensor.my_station_water_level_stage")
+    assert state.state == "5"
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5985897)